File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2017/shared/hashes.kt
package org.gristle.pdxpuzzles.advent.y2017.shared

import org.gristle.pdxpuzzles.utilities.objects.shift

fun denseHash(lengths: List<Int>): String {
    val ring = List(256) { it }
    val shiftSum = lengths.sum() * 64
    val skipSum = (lengths.size) * 64
    val p2 = (0 until 64).fold(ring) { acc, i ->
        acc.knotHash(lengths, i * (lengths.size))
    }
    val totalSkips = (1 until skipSum).reduce { acc, i -> acc + i }
    val reshifted = p2.shift(0 - (shiftSum + totalSkips))

    val denseHash = reshifted
        .chunked(16)
        .joinToString("") { chunk ->
            val reduction = chunk.reduce { acc, i -> acc xor i }
            String.format("%02x", reduction)
        }
    return denseHash
}

fun <E> List<E>.knotHash(lengths: List<Int>, skip: Int = 0): List<E> {
    return lengths.foldIndexed(this) { index, accRing, length ->
        knot(accRing, length, skip + index)
    }
}

private fun <E> knot(ring: List<E>, length: Int, skip: Int): List<E> {
    val reversePart = ring.subList(0, length).reversed()
    return (reversePart + ring.drop(reversePart.size)).shift(length + skip)
}

File: rust/advent/src/utilities/hashes.rs
use itertools::Itertools;

pub fn dense_hash(lengths: &[usize]) -> String {
    let ring: Vec<usize> = (0..256).collect();   
    let shift_sum = lengths.iter().sum::<usize>() * 64;
    let skip_sum = lengths.len() * 64;
    let mut reshifted = (0..64).fold(ring, |acc, i| knot_hash(acc, lengths, i * lengths.len()));
    let reshifted_len = reshifted.len();
    let total_skips = (1..skip_sum).sum::<usize>();
    reshifted.rotate_right((total_skips + shift_sum) % reshifted_len);

    let dense_hash = reshifted.into_iter().chunks(16).into_iter()
        .map(|chunk| {
            let reduction = chunk.reduce(|acc, i| acc ^ i).unwrap();
            format!("{:02x}", reduction)
        })
        .collect();
    dense_hash
}

pub fn knot_hash(ring: Vec<usize>, lengths: &[usize], skip: usize) -> Vec<usize> {
    lengths.iter().enumerate().fold(ring, |acc, (index, &length)| {
        let reverse_part: Vec<usize> = acc[..length].iter().rev().copied().collect();
        let mut knot: Vec<usize> = reverse_part.iter()
            .chain(acc.iter().skip(reverse_part.len()))
            .copied()
            .collect();
        let knot_len = knot.len();
        knot.rotate_left((length + skip + index) % knot_len);
        knot
    })
}
